src/main/java/quickcache/logic/commands/DeleteCommand.java

```diff
@@ -18,36 +22,98 @@
     public static final String COMMAND_WORD = "delete";
 
     public static final String MESSAGE_USAGE = COMMAND_WORD
-            + ": Deletes the flashcard identified by the index number used in the displayed flashcard list.\n"
-            + "Parameters: INDEX (must be a positive integer)\n"
-            + "Example: " + COMMAND_WORD + " 1";
+            + ": Deletes flashcards from flashcard list.\n"
+            + "If an INDEX is given, then the flashcard identified by the index number is deleted from the displayed flashcard list.\n"
+            + "If tag prefixes is given, the all flashcards with those tags will be deleted as well.\n"
+            + "An error will be thrown if both index and tag prefixes are given together.\n"
+            + "Parameters (either but not both): \n"
+            + "1. INDEX (must be a positive integer)\n"
+            + "2. [TAGS]\n"
+            + "Examples: \n"
+            + "1. " + COMMAND_WORD + " 1\n"
+            + "2. " + COMMAND_WORD + " " + PREFIX_TAG + "MCQ";
 
     public static final String MESSAGE_DELETE_FLASHCARD_SUCCESS = "Deleted Flashcard: %1$s";
 
     private final Index targetIndex;
+    private final FlashcardPredicate predicate;
+    private final Set<Tag> tagsToMatch;
+    private final boolean isDeleteByTag;
 
-    public DeleteCommand(Index targetIndex) {
+    private DeleteCommand(Index targetIndex) {
         this.targetIndex = targetIndex;
+        this.predicate = null;
+        this.tagsToMatch = null;
+        this.isDeleteByTag = false;
+    }
+
+    private DeleteCommand(FlashcardPredicate predicate, Set<Tag> tagsToMatch) {
+        this.targetIndex = null;
+        this.predicate = predicate;
+        this.tagsToMatch = tagsToMatch;
+        this.isDeleteByTag = true;
+    }
+
+    public static DeleteCommand withIndex(Index targetIndex) {
+        return new DeleteCommand(targetIndex);
+    }
+
+    public static DeleteCommand withPredicate(FlashcardPredicate predicate, Set<Tag> tagsToMatch) {
+        return new DeleteCommand(predicate, tagsToMatch);
+    }
+
+    private String createDeleteWithTagsMessage() {
+        StringBuilder sb = new StringBuilder("with tags ");
+        assert isDeleteByTag && tagsToMatch != null;
+        tagsToMatch.forEach(tag -> sb.append(tag.toString()).append(" "));
+        return sb.toString().trim();
```

Comment:
Perhaps you can make use of `.reduce()` to make this method even more concise.
```java
return tagsToMatch.stream().reduce(("with tags ", (curr, next) -> curr + next.toString() + " ").trim()
```

Reply:
truly a CS2030 TA, thanks for the suggestion!